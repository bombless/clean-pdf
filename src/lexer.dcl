definition module lexer

:: Token = StringLiteral [Char]
         | Key [Char]
         | DictStart
         | DictEnd
         | ListStart
         | ListEnd
         | StreamStart
         | StreamEnd
         | ObjectStart Int Int
         | ObjectEnd
         | Ref Int Int
         | Number Real
         | XRef
         | Eof
         | Id [Char]
         | Null
         | Bool Bool

parse :: [Char] -> [Token]

formatTokenList :: [Token] -> String
