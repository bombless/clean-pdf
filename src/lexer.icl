implementation module lexer

import StdEnv
import lexer

instance toTokens [Char] where
    toTokens :: [Char] -> [Token]
    toTokens _ = [Null]

instance toTokens {#Char} where
    toTokens :: {#Char} -> [Token]
    toTokens x = parseFromList (fromString x)
    where
        parseFromList :: [Char] -> [Token]
        parseFromList x = toTokens x

instance toString Token where
    toString (StringLiteral _) = "StringLiteral"
    toString (Key _)           = "Key"
    toString DictStart         = "DictStart"
    toString DictEnd           = "DictEnd"
    toString ListStart         = "ListStart"
    toString ListEnd           = "ListEnd"
    toString StreamStart       = "StreamStart"
    toString StreamEnd         = "StreamEnd"
    toString (ObjectStart _ _) = "ObjectStart"
    toString ObjectEnd         = "ObjectEnd"
    toString (Ref _ _)         = "Ref"
    toString (Number _)        = "Number"
    toString XRef              = "XRef"
    toString Eof               = "Eof"
    toString (Id _)            = "Id"
    toString Null              = "Null"
    toString (Bool b)          = "Bool(" +++ toString b +++ ")"

formatTokenList :: [Token] -> String
formatTokenList x = "[" +++ formatTokenListHelper x +++ "]"
where
    formatTokenListHelper :: [Token] -> String
    formatTokenListHelper [] = ""
    formatTokenListHelper [x] = toString x
    formatTokenListHelper [x:xs] = toString x +++ ", " +++ formatTokenListHelper xs

instance == Token where
    (==) (StringLiteral a) (StringLiteral b) = a == b
    (==) (Key a) (Key b)                     = a == b
    (==) Null Null                           = True
    (==) (Number a) (Number b)               = a == b
    (==) _ _                                 = False

