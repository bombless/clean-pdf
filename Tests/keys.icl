module keys

import lexer
import StdEnv
import helper

Start = (formatTokenList o toTokens) "/1 2 /3 4"
