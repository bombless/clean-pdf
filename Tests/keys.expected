[/1, Number(2), /3, Number(4)]
